highfidelity/hifi
# 21458 - Fix Shortbow

Fix Shortbow
- Created didRootPositionMove and checkRootPosition in
shortbowGameManager to handle updating spawn positions if board is moved
during game play

- Created for functions for creating and deleting bows and enemy spawn
positions

- Fixed lives check

- Fixed error causing pieces to fall off board.  This was being caused
by a jump in velocity during a random time which was interfering with
where the entity was being redrawn

- Added a try/catch in handControllerGrab.js to fix problem with trying
to get properties from cache that were being deleted when too many
arrows at one time were being thrown.

Test Plan : 
For Section 1 :
-------------------------------------------------------------
Re:
1) Laser/stylus become unusable
Steps to reproduce:
1) Get the bow from the marketplace
2) When you first get it, the laser/stylus works, you can drop it and use the grip or trigger to pick it up with either hand
3) Very rapidly shoot multiple arrows (I think I did around 10) in the same general direction--I didn't even bother pulling back all the way
-------------------------------------------------------------
- If you are able to shoot multiple arrows as fast as you can shoot, then this problem should be fixed. 

My computer was giving me issues at first around 7-10 but the number would change depending on system resources available. 

For Section 2:
-------------------------------------------------------------
Re:

2) Stops spawning balls and bows
Steps to reproduce:
1) Go to marketplace and add the shortbow game to your domain
2) Hit play
3) Enter Create/edit mode and move the game 

The spawn points don't move with the game. For some shortbow games the balls spawn, behind the avatar shop instead of in the pit and will freely roll anywhere in the domain. Since the game itself is not a locked asset, this is probably from it getting moved after being placed the first time. However, no balls spawned and the start sign failed to appear after wave 10, so it may be losing the ability to keep track of how many balls are still present.
-------------------------------------------------------------

- run spawnShortBow.js
- Most of this problem had to do with the randomized jump on the enemy balls after being initiated too quickly after spawning.  There seems to be a general problem related to disappearing models when Entities.getEntityProperties get called. 
- Moving the board around while playing should reset the wave and spawn positions
- No balls should be outside of the play area
- Lives should be tracking fine despite the wave number.
